packages/ipfs-core/src/utils/service.js

```diff
@@ -0,0 +1,239 @@
+          return result
+        // If failed to start, transiton from 'starting' to 'stopped'
+        // state.
+        } catch (error) {
+          service.state = { status: 'stopped' }
+          throw error
+        }
+      }
+      case 'starting': {
+        throw new AlreadyStartingError()
+      }
+      case 'started': {
+        throw new AlreadyStartedError()
+      }
+      // If service is stopping we just wait for that to complete
+      // and try again.
+      case 'stopping': {
+        await state.ready
+        return await Service.start(service, options)
+      }
+      default: {
+        return Service.panic(service)
+      }
+    }
+  }
+
+  /**
+   * Stops the service by executing deactivation. If service is stopped
+   * or is stopping this is noop. If service is starting up when called
+   * it will await for start to complete and then retry stop afterwards.
+   * This may (async) throw if `deactivate` does.
+   *
+   * @template T
+   * @param {Service<any, T>} service
+   * @returns {Promise<void>}
+   */
+  static async stop (service) {
+    const { state, deactivate } = service
+    switch (state.status) {
+      // If stopped there's nothing to do.
+      case 'stopped': {
+        break
+      }
+      // If service is starting we await for it to complete
+      // and try again. That way
+      case 'starting': {
+        // We do not want to error stop if start failed.
+        try { await state.ready } catch (_) {}
+        return await Service.stop(service)
+      }
+      // if service is stopping we just await for it to complete.
+      case 'stopping': {
+        return await state.ready
+      }
+      case 'started': {
+        if (deactivate) {
+          await deactivate(state.value)
+        }
+        service.state = { status: 'stopped' }
+        break
+      }
+      default: {
+        Service.panic(state)
+      }
+    }
+  }
+
+  /**
+   * @template T
+   * @param {Service<any, T>} service
+   * @returns {T|null}
+   */
+  static try ({ state }) {
+    switch (state.status) {
+      case 'started':
+        return state.value
+      default:
+        return null
+    }
+  }
+
+  /**
+   * Unwraps state and returns underlying value. If state is in idle state it
+   * will throw an error. If state is pending it will wait and return the
+   * result or throw on failure. If state is ready returns result.
+   *
+   * @template T
+   * @param {Service<any, T>} service
+   * @param {AbortOptions} [options]
+   * @returns {Promise<T>}
+   */
+  static async use ({ state }, options) {
+    switch (state.status) {
+      case 'started':
+        return state.value
+      case 'starting':
+        return await withTimeout(state.ready, options)
+      default:
+        throw new NotStartedError()
+    }
+  }
+
+  // eslint-disable-next-line jsdoc/require-returns-check
+  /**
+   * @private
+   * @param {Service<any, any>} service
+   * @returns {never}
```

Comment:
Can this be `@returns {void}`?  Then apparently we don't need the `eslint-disable-next-line` - https://eslint.org/docs/rules/valid-jsdoc

Reply:
No it needs to be `never` because it can never return and by consequence can be used e.g. to here:

```ts
const div = (a:number, b:number) : number => {
  if (b === 0) return panic() // never
  return a / b
}
```

Note that if `panic` returned `void` type checker would complain that return type is `number|void` and not a `number`. While`never|number` is equivalent of `number`.

Here specifically it used in switch statements